### DISCUSSION TITLE
System.Drawing.Graphics.FillPie() analog

### DISCUSSION THREAD

**Discussion by luber** [SixLabors/ImageSharp.Drawing#213] [Q&A]
Hi guys,
I am trying to replace System.Drawing.Graphics.FillPie method with ImageSharp.
Could you please guide me on how can I achieve this?
I need to draw a small pie chart with 3 dynamic colours on the icon. Something like:
![image](https://user-images.githubusercontent.com/3791692/160660297-2fb0eda7-d318-4354-b5d3-294e794100ea.png)

**Comment by tocsoft** [comment]
We don't have a Pie segment as a built-in primitive but you should be able to use the `PathBuilder` class to create any complex shape you require.

`PathBuilder`  allows you to directively 'draw' your pie segment, using a series of lines and curves. You then take the output of the builder and feed that along into the Fill APIs with what ever color you want.

**Reply by luber** [reply]
Thanks for the quick answer. I have tried to do this like that:
```
using (Image img = new Image<Rgba32>(width, height))
{
    DrawingOptions options = new()
    {
        GraphicsOptions = new()
        {
            ColorBlendingMode  = PixelColorBlendingMode.Multiply
        }
    };
    
    var imgCenter = new Point(width / 2, height / 2);
    var radius = width / 2;

    IBrush brush = Brushes.Solid(backColor);
    IPen pen = Pens.Solid(Color.Black, 1);
    
    IPath path = isRectangle
        ? new RectangularPolygon(1, 1, img.Width - 2, img.Height - 2)
        : new EllipsePolygon(imgCenter, img.Width / 2 - 2);
    
    img.Mutate(ctx => ctx
        .Fill(options, brush, path)
        .Draw(options, pen, path)
        .Fill(Brushes.Solid(Color.Red),
            pb =>
            {
                pb.AddLine(imgCenter, new PointF(imgCenter.X - radius, imgCenter.Y - radius));
                pb.AddLine(imgCenter, new PointF(imgCenter.X + radius, imgCenter.Y - radius));
                pb.AddEllipticalArc(imgCenter, radius, radius, 0, 180, 45);
            })                
    );
    
    img.SaveAsPng("D:\\test.png");
}

```

But it produces a bit not expected result:
![test](https://user-images.githubusercontent.com/3791692/160681081-e3519bc1-4bcf-4345-a109-76a4e4ec60aa.png)

Any ideas about what am I doing wrong?

**Reply by luber** [reply]
Looks like I get it. You need to draw polygon by "walking" from point to point. So the order matters:
```
 .Fill(Brushes.Solid(Color.Red),
                        pb =>
                        {
                            var endL1 = new PointF(imgCenter.X - radius, imgCenter.Y - radius);
                            
                            pb
                                .AddLine(imgCenter, endL1)
                                .AddLine(endL1, imgCenter)
                                .AddEllipticalArc(imgCenter, radius, radius, 0, 135, 90);
                        })
```

Not image looks correct:
![test](https://user-images.githubusercontent.com/3791692/160686159-9c7f5f30-7954-4366-93ff-d4ad7bf62feb.png)
